design/authenticators/authn_sut/authn_sut_solution_design.md

```diff
@@ -0,0 +1,384 @@
+enough time to allow the client to use the SUT to authenticate to Conjur, even
+in the case of network latency, but short enough to limit the window of
+opportunity for an attacker to steal and use the SUT.
+
+Additionally, only one SUT will be valid at a time for a given user. If a new
+SUT is generated for a user, the previous SUT will be deleted.
+
+When a SUT is used (or revoked), it will be deleted from the database.
+Expired SUTs can also safely be deleted from the database, though
+this is not strictly necessary since they will not be valid past
+their expiration time.
+When deleting used or expired SUTs, we must not cascade the delete
+since this would delete any related audit logs.
+
+##### Code Verifier and Code Challenge
+
+As an additional layer of security, the client will generate a code verifier,
+which can be a Guid or other cryptographically secure random string. The code
+will then be hashed and sent along with the login request to be associated
+with the generated SUT. This hashed code verifier, known as the code
+challenge, will be stored by Conjur along with the SUT. When the client
+attempts to authenticate with the SUT, they will also send the code verifier
+(in plain text form), and Conjur will hash it and compare it to the stored
+code challenge.
+
+Initially we will support the SHA-256 algorithm for the code challenge. In the
+future we may want to support additional algorithms.
+
+The code challenge and code verifier flow are modeled after OIDC PKCE.
+
+##### Open Questions
+
+- Is it possible to restrict use of the SUT to a specific IP range (like with
+  hostfactory tokens) or other criteria to limit risk if the SUT is leaked?
+- Should we provide a way for clients to register a key pair to use for
+  signing the SUT generation request? This would allow us to verify that the
+  request came from a trusted client.
+- Should we create a job to clean up expired SUTs?
+- Should we create a mechanism to revoke SUTs before they expire (other than
+  manually deleting them from the database)?
+
+##### Example Database Table
+
+| id | role_id | guid_hash | code_challenge | code_challenge_algorithm | expires_at |
+|----|---------|-----------|----------------|--------------------------|------------|
+| 1  | user:john.doe | lkduj...  | 3iod... | sha256 | 2023-09-28 12:08:23 |
+| 2  | user:demo.user | 2k1g2...  | kld80... | sha256 | 2023-09-28 12:09:45 |
+
+For this example, we would need a migration to add the `authn_sut_tokens` table.
+That would look something like this:
+
+```ruby
+Sequel.migration do
+  change do
+    create_table :authn_sut_tokens do
+      primary_key :id
+      foreign_key :role_id, :roles, type: String, null: false, on_delete: :cascade
+      String :guid_hash, null: false, unique: true
+      String :code_challenge, null: false
+      String :code_challenge_algorithm, null: false
+      DateTime :expires_at, null: false
+    end
```

Comment:
This table is missing an index.  I'd recommend creating an index on the on `guid_hash` and `expires_at`.  This will allow us to run something like this with a level of efficiency:

```sql
select role_id, code_challenge, code_challenge_algorithm from authn_sut_tokens where guid_hash = '...' and expires_at >= NOW();
```